Build binaries inside Docker

bandesz opened the pull request:
## What

It's a good common practice to build our binaries inside Docker, using a multistage build. This way we can make sure the binary is built in the same environment where it's going to run.

This will help to build the Docker images locally if required.

This change will need vendoring as we don't want to download the dependencies for every build.

lewismarshall commented on images/Dockerfile.kore-apiserver:
```diff
@@ -7,14 +16,13 @@ LABEL Name=kore-apiserver \
 
-COPY bin/kore-apiserver /kore-apiserver
-COPY bin/kore-clusterappman /kore-clusterappman
+COPY --from=0 /go/src/github.com/appvia/kore/bin/* /bin/
```
If you change where these are we need to update the kore-clusterappman manifest as it currently uses an exec, see: https://github.com/appvia/kore/blob/b145e5a3d74f6b8ecf2717e0c67fc7f7c6afd3cd/pkg/controllers/management/kubernetes/helper.go#L308

lewismarshall replied:
@bandesz so added this commit on a branch if you want to `cherry-pick` or otherwise include in your PR:
https://github.com/appvia/kore/commit/3b92af90b70e10fc62ee995d8a6c10410ba15ce4

lewismarshall replied:
@bandesz so added this commit on a branch if you want to `cherry-pick` or otherwise include in your PR:
https://github.com/appvia/kore/commit/3b92af90b70e10fc62ee995d8a6c10410ba15ce4

lewismarshall replied:
@bandesz so added this commit on a branch if you want to `cherry-pick` or otherwise include in your PR:
https://github.com/appvia/kore/commit/3b92af90b70e10fc62ee995d8a6c10410ba15ce4

bandesz replied:
Fixed

bandesz commented:
@lewismarshall thanks for the review, I removed the build dependency from the docker targets. Also I removed the static target as it doesn't have a purpose anymore.

lewismarshall commented on images/Dockerfile.auth-proxy:
```diff
@@ -1,14 +1,25 @@
+RUN apk --no-cache add ca-certificates
 
-COPY bin/auth-proxy /auth-proxy
+COPY --from=0 /go/src/github.com/appvia/kore/bin/* /bin/
```
This should probably just be `auth-proxy` only if we're going to the pain of having a seperate image?

bandesz replied:
We build only the auth-proxy command (the bin directory is excluded in .dockerignore), so we can use * here.

bandesz commented:
@lewismarshall I addressed your comments, are you okay to leave the additional CircleCI work for a future PR?

lewismarshall commented:
> @lewismarshall I addressed your comments, are you okay to leave the additional CircleCI work for a future PR?

I think we need all the signals we can get to help in the review process and not having CI working is probably not ever a good idea so can we get another commit on this PR to fix Circle CI, will probs simplify the whole build just a `docker build` on the build stage and then a `docker push` required and no artifacts between built and push stages.

bandesz commented:
@lewismarshall there is no cross-compilation risk, so I don't think we need additional CircleCI work right now.
